src/adapter/bridges/OFTBridge.ts

```
@@ -0,0 +1,318 @@
+   * @param l1Token - Token address on L1
+   * @param _l2Token - Token address on L2 (not used, determined by bridge)
+   * @param amount - Amount to transfer
+   * @returns Transaction details for execution
+   */
+  async constructL1ToL2Txn(
+    toAddress: string,
+    l1Token: string,
+    _l2Token: string,
+    amount: BigNumber
+  ): Promise<BridgeTransactionDetails> {
+    // Verify the token matches the one this bridge was constructed for
+    if (l1Token !== this.hubTokenAddress) {
+      throw new Error(`This bridge instance only supports token ${this.hubTokenAddress}, not ${l1Token}`);
+    }
+
+    // SAFETY LIMIT: Cap transfer amount to prevent large transfers
+    // This is retained as a safety measure to protect against large accidental transfers
+    if (amount.gt(OFTBridge.MAX_AMOUNT)) {
+      throw new Error(`Amount exceeds maximum allowed (${amount} > ${OFTBridge.MAX_AMOUNT})`);
+    }
+
+    // we need to be careful what amounts we pass into OFT transfers to prevent rounding
+    await this.validateAmountPrecision(amount);
+
+    // Construct the send parameters for the OFT bridge
+    // @dev last `(supportedToken.decimals() - IOFT.sharedDecimals())` digits in amount
+    // must be zero to prevent rounding on the contract side
+    const sendParamStruct: SendParamStruct = {
+      dstEid: this.dstChainEid,
+      to: oftAddressToBytes32(toAddress).toLowerCase(),
+      amountLD: amount,
+      minAmountLD: amount,
+      extraOptions: "0x", // Empty bytes
+      composeMsg: "0x", // Empty bytes
+      oftCmd: "0x", // Empty bytes
+    };
+
+    // Get the messaging fee for this transfer
+    const l1Bridge = this.l1Bridge;
+    const feeStruct: MessagingFeeStruct = await (l1Bridge as IOFT).quoteSend(sendParamStruct, false);
+
+    if (BigNumber.from(feeStruct.nativeFee).gt(OFTBridge.FEE_CAP)) {
+      throw new Error(`Fee exceeds maximum allowed (${feeStruct.nativeFee} > ${OFTBridge.FEE_CAP})`);
+    }
```

If we ever hit this in production then it'll take the rebalancer down and will prevent bridging to any other chains. That would require a code change, or at least a config change. I think it makes sense to instead alert about it via a log messsage and abort the deposit.

@bmzig We don't have a logger instance in this class; do you think there could be a case for adding it to the `BaseBridgeAdapter` ?

---

Yeah, I'll remove the `throw` here. As for aborting: in general, what's our policy on returning errors? If I return a `null`, will the caller handle it fine? Or how do you suggest we cancel the tx here?

---

At the moment we're forced to return a `BridgeTransactionDetails` type, which doesn't really allow for failure.

In some places we return a union type of (<success>|<failure>) (i.e. in the SpokePoolClient [here](https://github.com/across-protocol/sdk/blob/9bce8dd8ca80587b2f7a65a6472d0fe763f3ee19/src/clients/SpokePoolClient/SpokePoolClient.ts#L58)). In practice this works well if there are various types of sometimes-expected failures that can happen.

This seems more like a "should never ever happen" type thing, so the simplest is probably to keep this exception, and then add a try/catch clause [here](https://github.com/across-protocol/relayer/blob/02a666f39dd426b3e4422b95b97a316fd6430d58/src/adapter/BaseChainAdapter.ts#L236).

---

Looking at this once more, in `sendTokenToTargetChain` that you link to (for adding a try/catch, we actually throw on [e.g. simulation failure](https://github.com/across-protocol/relayer/blob/02a666f39dd426b3e4422b95b97a316fd6430d58/src/adapter/BaseChainAdapter.ts#L252). So if we throw from `constructL1ToL2Txn` if the params are incorrect it should be fine too.

The only extra thing that `try / catch` would allow us to do is to log, as like you mentioned there's no logger in `BaseBridgeAdapter`.

So we could either:

1. Leave throw as is, let some caller catch it (which should happen if `sendTokenToTargetChain` throws tell us anything)
2. Try - catch + log + re-throw

Which one do you prefer we do? @pxrl @bmzig

---

Good point about the fact that we can already throw errors from within `sendTokenToTargetChain()`. I think in practice this almost never throws, because it looks like it would take down the bot (I don't see any error suppression upstream in the `AdapterManager` or `InventoryClient` classes). So the fact that we don't suppress the error doesn't necessarily mean we would handle such an error gracefully.

I do still think it's important to ensure that the failure to deposit a single token doesn't result in contagion preventing any other tokens from being rebalanced. The `InventoryClient` class expects to receive a `TransactionResponse`  promise but it only retains the `hash`, so that might be an OK place to log and suppress the failure without having to re-jig types all over the place.

---

@pxrl [581865f](https://github.com/across-protocol/relayer/pull/2133/commits/581865fc03e352e64bab19a3e61027c900d00984) is this what you had in mind?

---

Yeah; this is OK. It would blow up if the caller calls `TransactionResponse.wait()` on a failed response, but we don't do that in practice, and if we do it'd imply a larger refactoring anyway.